File: backend/app/api/tribulation.py
"""渡劫 HTTP 路由（M5）。"""

from __future__ import annotations

from fastapi import APIRouter, Depends

from app.core.deps import get_current_user, get_tribulation_service
from app.db.models import User
from app.schemas.common import success
from app.schemas.tribulation import TribulationPrepRequest, TribulationResolveBatchRequest
from app.services.tribulation_service import TribulationService

router = APIRouter(prefix="/tribulation", tags=["tribulation"])


@router.get("/me", response_model=None)
async def tribulation_me(
    svc: TribulationService = Depends(get_tribulation_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """当前渡劫会话。"""
    return success(await svc.get_me(current_user))


@router.post("/start-prep", response_model=None)
async def tribulation_start_prep(
    svc: TribulationService = Depends(get_tribulation_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """创建 preparing 会话。"""
    return success(await svc.start_prep(current_user))


@router.put("/prep", response_model=None)
async def tribulation_save_prep(
    payload: TribulationPrepRequest,
    svc: TribulationService = Depends(get_tribulation_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """保存准备格。"""
    return success(
        await svc.save_prep(
            current_user,
            slots=payload.slots,
            formation_id=payload.formation_id,
            veil_chosen=payload.veil_chosen,
        ),
    )


@router.post("/commit-prep", response_model=None)
async def tribulation_commit_prep(
    svc: TribulationService = Depends(get_tribulation_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """确认准备。"""
    return success(await svc.commit_prep(current_user))


@router.post("/veil-check", response_model=None)
async def tribulation_veil_check(
    svc: TribulationService = Depends(get_tribulation_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """遮天检定。"""
    return success(await svc.veil_check(current_user))


@router.post("/begin", response_model=None)
async def tribulation_begin(
    svc: TribulationService = Depends(get_tribulation_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """开渡。"""
    return success(await svc.begin(current_user))


@router.post("/resolve-batch", response_model=None)
async def tribulation_resolve_batch(
    payload: TribulationResolveBatchRequest | None = None,
    svc: TribulationService = Depends(get_tribulation_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """结算下一批雷击。"""
    batch_size = payload.batch_size if payload else None
    return success(await svc.resolve_batch(current_user, batch_size=batch_size))


@router.post("/auto-resolve", response_model=None)
async def tribulation_auto_resolve(
    svc: TribulationService = Depends(get_tribulation_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """一键结算至结束。"""
    return success(await svc.auto_resolve(current_user))

File: frontend/src/api/tribulation.ts
/**
 * M5 渡劫 API：准备 / 遮天 / 开渡 / 批次结算。
 */
import { http } from './http'
import { envelopeFromAxiosError } from './envelope'
import type { ApiResponse } from '../types/api'
import type {
  TribulationMutationResult,
  TribulationPrepPayload,
  TribulationSessionPublic,
  TribulationVeilResult,
} from '../types/tribulation'

/** GET /tribulation/me */
export async function fetchTribulation(): Promise<
  ApiResponse<{ session: TribulationSessionPublic | null }>
> {
  try {
    const response = await http.get<
      ApiResponse<{ session: TribulationSessionPublic | null }>
    >('/tribulation/me')
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<{ session: TribulationSessionPublic | null }>(error)
  }
}

/** POST /tribulation/start-prep */
export async function startPrep(): Promise<ApiResponse<TribulationMutationResult>> {
  try {
    const response = await http.post<ApiResponse<TribulationMutationResult>>(
      '/tribulation/start-prep',
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<TribulationMutationResult>(error)
  }
}

/** PUT /tribulation/prep */
export async function savePrep(
  payload: TribulationPrepPayload,
): Promise<ApiResponse<TribulationMutationResult>> {
  try {
    const response = await http.put<ApiResponse<TribulationMutationResult>>(
      '/tribulation/prep',
      payload,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<TribulationMutationResult>(error)
  }
}

/** POST /tribulation/commit-prep */
export async function commitPrep(): Promise<ApiResponse<TribulationMutationResult>> {
  try {
    const response = await http.post<ApiResponse<TribulationMutationResult>>(
      '/tribulation/commit-prep',
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<TribulationMutationResult>(error)
  }
}

/** POST /tribulation/veil-check */
export async function veilCheck(): Promise<ApiResponse<TribulationVeilResult>> {
  try {
    const response = await http.post<ApiResponse<TribulationVeilResult>>(
      '/tribulation/veil-check',
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<TribulationVeilResult>(error)
  }
}

/** POST /tribulation/begin */
export async function beginTribulation(): Promise<
  ApiResponse<TribulationMutationResult>
> {
  try {
    const response = await http.post<ApiResponse<TribulationMutationResult>>(
      '/tribulation/begin',
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<TribulationMutationResult>(error)
  }
}

/** POST /tribulation/resolve-batch */
export async function resolveBatch(body?: {
  batch_size?: number
}): Promise<ApiResponse<TribulationMutationResult>> {
  try {
    const response = await http.post<ApiResponse<TribulationMutationResult>>(
      '/tribulation/resolve-batch',
      body ?? {},
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<TribulationMutationResult>(error)
  }
}

/** POST /tribulation/auto-resolve */
export async function autoResolve(): Promise<ApiResponse<TribulationMutationResult>> {
  try {
    const response = await http.post<ApiResponse<TribulationMutationResult>>(
      '/tribulation/auto-resolve',
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<TribulationMutationResult>(error)
  }
}
